### PR TITLE
basicauth: Add a allow_cidr option to whitelist IPs

### PR DESCRIPTION
**Pull request by gregdel**
### 1. What does this change do, exactly?

This PR adds a new parameter: `allowed_cidr` for the basicauth module. It allows bypassing the basic auth based on the remote ip of the request.

### 2. Please link to the relevant issues.

https://github.com/mholt/caddy/issues/1891

### 3. Which documentation changes (if any) need to be made because of this PR?

The documentation of the basicauth module needs to be updated if this PR is merged.

### 4. Checklist

- [x] I have written tests and verified that they fail without my change
- [x] I have squashed any insignificant commits
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I am willing to help maintain this change if there are issues with it later

### CODE REVIEW

**Comment by mholt** [comment]
Neat -- have you seen the `ipfilter` plugin? https://caddyserver.com/docs/http.ipfilter - does this do something different than that?

**Comment by gregdel** [comment]
From my understanding of the ipfilter package, you can allow or block ips for a vhost.
My use case is to replace this configuration from nginx:
```
location / {
    satisfy any;
    allow x.x.x.x;
    allow x.x.x.x;

    deny all;
    auth_basic "Login required";
    auth_basic_user_file /somefile/containing/the/users;

    autoindex on;
    root /some/directory/containing/files;
}
```

Everybody gets the basic auth except for the authorized ips.

I didn't manage to do this with ipfilter but I might not understand it correctly.

**Comment by whalehub** [comment]
@mholt @gregdel I've tested this use case just now and can confirm that the ipfilter package does **not** allow you to do this. When you combine the basic auth directive with the ipfilter directive, a host whose IP address is not listed in the whitelist receives a 401 status code when accessing the page while a host whose IP address is listed on the whitelist receives the basic auth prompt.

The purpose of this pull request, from what I understand, is to allow anyone, regardless of their IP address, to receive the basic auth prompt while allowing those on the whitelist to bypass it entirely. I hope that clears things up so that this (very useful!) PR can be merged in to the master branch ASAP.

**Comment by whalehub** [comment]
@mholt @gregdel I've tested this use case just now and can confirm that the `ipfilter` plugin does **not** allow you to do replace that NGINX configuration.

When you combine the basic auth directive with the ipfilter directive in your Caddyfile,, a host whose IP address is not listed on the whitelist receives a 401 status code (i.e. does not receive the basic auth prompt) when accessing the page while a host whose IP address is listed on the whitelist does receive the basic auth prompt.

The purpose of this pull request, from what I understand, is to allow anyone, regardless of their IP address, to receive the basic auth prompt while allowing only those whose IP address is listed on the whitelist to bypass it (i.e. to access the page without entering their credentials).

I hope that this helps clear things up so that this awesome PR can be merged into the master branch ASAP.

**Comment by whalehub** [comment]
@mholt @gregdel I've tested this use case just now and can confirm that the ipfilter plugin does not allow you to replace that NGINX configuration.

When you combine the basic auth directive with the ipfilter directive in your Caddyfile,, a host whose IP address is not listed on the whitelist receives a 401 status code (i.e. does not receive the basic auth prompt) when accessing the page while a host whose IP address is listed on the whitelist does receive the basic auth prompt.

The purpose of this pull request, from what I understand, is to allow anyone, regardless of their IP address, to receive the basic auth prompt while allowing only those whose IP address is listed on the whitelist to bypass it (i.e. to access the page without entering their credentials).

I hope that this helps clear things up so that this awesome PR can be merged into the master branch ASAP.

**Comment by whalehub** [comment]
@mholt @gregdel I've tested this use case just now and can confirm that the ipfilter plugin does not allow you to replace that NGINX configuration.

When you combine the basic auth directive with the ipfilter directive in your Caddyfile, a host whose IP address is not listed on the whitelist receives a 401 HTTP status code (i.e. does not receive the basic auth prompt) when accessing the page while a host whose IP address is listed on the whitelist does receive the basic auth prompt.

The purpose of this pull request, from what I understand, is to allow anyone, regardless of their IP address, to receive the basic auth prompt while allowing only those whose IP address is listed on the whitelist to bypass it (i.e. to access the page without entering their credentials).

I hope that this helps clear things up so that this awesome PR can be merged into the master branch ASAP.

**Comment by whalehub** [comment]
@mholt Is there a reason why this PR hasn't been merged yet?

**Comment by tobya** [comment]
Thanks for the PR.  Would you be able to provide basic documentation changes that would be added to the docs for basicauth

**Comment by mholt** [comment]
Sorry, going to be busy for a couple months finishing my thesis.

**Comment by mholt** [comment]
Do we need to consider the role of IP spoofing with regards to this change?

**Comment by mholt** [comment]
Right there in the original post for the PR is fine, just as the template asks for. Saying "changes to the basicauth page" isn't helpful. _Which_ changes??

**Comment by gregdel** [comment]
Here's what we could add to the documentation:

#### Syntax section
`allowed_cidr` identifies IP/range of IPs for which to bypass the basic authentication. It accepts IPv4 and IPv6, a single IP or a range of IPs in CIDR format. This parameter can be repeted multiple times, one CIDR per line.

#### Exemple section
We could add this in the exemple:
```
basicauth "Mary Lou" milkshakes {
    realm "Mary Lou's documents"
    /notes-for-mary-lou.txt
    /marylou-files
    /another-file.txt
    allowed_cidr 127.0.0.1
    allowed_cidr ::1
    allowed_cidr 10.0.0.0/16
}
```
This shows IPv4/IPv6, IP and CIDR format. What do you think ?

**Comment by mholt** [comment]
Thanks! That's great.

As for IP spoofing, I don't know how to solve it, off-hand. Does this change assume that IP spoofing is not in the threat model? If so, we could document that.

**Comment by gregdel** [comment]
I think IP spoofing is not a real threat here and basic auth is not a strong security. The user should be aware of the security implication when using this option.

**Comment by whalehub** [comment]
@gregdel Agreed. I think this PR can be merged as-is @mholt.

**Comment by mholt** [comment]
Wouldn't a stronger implementation of this feature be to restrict basicauth to only certain IPs? In other words, still expect authentication, but don't allow other IPs to authenticate, even if they have the credentials. At least in this way, the allowed IPs can be kept secret, making spoofing harder, and you still have actual credentials protecting the resources.

Done correctly, basicauth isn't _weak_ per-se.

**Comment by gregdel** [comment]
The goal of this PR is to disable basic authentication entirely for some IPs.

**Comment by mholt** [comment]
Well, I don't love this idea, but we can go ahead and try it; noting that a plain IP address whitelist is not as assuring as a whitelist for authentication.

Thanks for the contribution -- hopefully you can find it useful, but again, I do hope that people will threat model properly and use this feature accordingly!

**Comment by mholt** [comment]
Specifically, if the attacker is "on-path" they can spoof the IP address for the complete connection -- the threat model must accept this if using CIDR whitelisting.

File: caddyhttp/basicauth/setup.go [hunk 1]
```diff
@@ -81,17 +82,54 @@ func basicAuthParse(c *caddy.Controller) ([]Rule, error) {
...
+				switch val {
+				case "realm":
 					if rule.Realm == "" {
 						rule.Realm = strings.Replace(args[0], `"`, `\"`, -1)
 					} else {
 						return rules, c.Errf("\"realm\" subdirective can only be specified once")
 					}
-				} else {
-					return rules, c.Errf("expecting \"realm\", got \"%s\"", val)
+				case "allowed_cidr":
```
Comment (mholt) [inline caddyhttp/basicauth/setup.go]: Let's change this to `insecure_allowed_cidr` in keeping with the proxy directive's `insecure_skip_verify`. At least this way people will know that what they're doing is insecure.

Comment (mholt) [reply caddyhttp/basicauth/setup.go]: Even better: `insecure_skip_auth_cidr` to make it clear that this doesn't just allow those IPs to authenticate, it skips authentication for those IPs.

Comment (mholt) [reply caddyhttp/basicauth/setup.go]: This is still a really bad idea IMO.

Why not just bind to a different interface and let your firewall do the work? This seems like a solution being implemented in the wrong place.

Comment (gregdel) [reply caddyhttp/basicauth/setup.go]: Don't merge it if you don't feel confortable about it, I won't be offended :)
This is only a proposal.

My use-case is to access mysite.com, if I'm at home, there's no security, if I'm at work I'll have to login. This is only for personal, non critical usage. If I'm the only one with this need, it should not be merged upstream.

This is however possible to configure nginx for this precise use-case:
```
location / {
    satisfy any;
    allow x.x.x.x;
    allow x.x.x.x;

    deny all;
    auth_basic "Login required";
    auth_basic_user_file /somefile/containing/the/users;

    autoindex on;
    root /some/directory/containing/files;
}
```

Comment (whalehub) [reply caddyhttp/basicauth/setup.go]: @mholt @gregdel I'm also interested in this use-case, so that makes two of us.

**Comment by magikstm** [comment]
I'm not in favor of this PR.

I think it adds too much extra complexity to the core of Caddy that would have to be maintained.

I think it would be better as a plugin.

**Comment by mholt** [comment]
K, yeah... I agree. This is not an ideal solution. Thanks for the PR, but I am going to close this.

**Pull request closed without merging**